File: kmeans.py
import math
import sys

iterations = 200
epsilon = 0.001

# read system args and check if valid
def get_sys_args():
    global iterations
    global k
    if len(sys.argv) == 1:
        print("An Error Has Occurred")
        sys.exit()
    # when len(sys.argv) > 1:
    try:
        k = int(sys.argv[1])
    except ValueError:
        try:
            k = float(sys.argv[1])
            print("Invalid number of clusters!")
            sys.exit()
        except ValueError:
            print("An Error Has Occurred")
            sys.exit()
    if len(sys.argv) == 3:
        file_name = sys.argv[2]
    elif len(sys.argv) == 4:
        try:
            iterations = int(sys.argv[2]) # iter is the second arg
            file_name = sys.argv[3]
        except ValueError:
            try:
                k = float(sys.argv[2])
                print("Invalid maximum iteration!")
                sys.exit()
            except ValueError:
                print("An Error Has Occurred")
                sys.exit()
    return file_name

# Open the file in read mode and load data to points
def read_file(file_name, points):
    with open(file_name, 'r') as file:
        file_contents = file.read()
        vectors = file_contents.split('\n')
        for i in range(len(vectors)-1):
            vector = vectors[i].split(',')
            for j in range(len(vector)):
                vector[j] = float(vector[j])
            points.append(vector)
    return

def create_centroids(centroids, points, k, d):
    # create centroids:
    for i in range(k):
        point = []
        for j in range(d):
            point.append(points[i][j])
        centroids.append(point)
    return centroids, points

def print_2d_mat(points):
    #print 2d matrix:
    for i in range(len(points)):
        for j in range(len(points[0])):
            print("%.4f" %points[i][j], end='')
            if j < len(points[0]) - 1:
                print(",", end='')
        print()


def kmeans(points, centroids, n, k, d):
    '''Implement k-means clustering algorithm'''
    
    for i in range(iterations):
        a_i_arr = assign(points, centroids, n, k, d) # a_i_dict := {cluster index : [point indices]}
        delta_c = update(a_i_arr, points, centroids, k, d) # delta_c := maximum distance between the prev. and cur. centroids 
        if convergence(delta_c):
            break


def assign(points, centroids, n, k, d):
    '''Assign every point to the closest cluster'''

    a_i_arr = [[] for i in range(k)]
    for i in range(n):
        min_index = argmin_distance(centroids,points[i], k, d)
        a_i_arr[min_index].append(i) 
    return a_i_arr


def update(a_i_arr, points, centroids, k, d):
    '''Update the centroids' coordinates to be the center of mass of their associated points'''

    delta_c = 0
    for i in range(k):
        length = len(a_i_arr[i])
        if length!=0:
            new_c = [sum([points[a_i_arr[i][j]][m] for j in range(length)])/length for m in range(d)]
            delta_c = max(distance_pp(new_c,centroids[i], d),delta_c)
            centroids[i] = new_c
    return delta_c


def convergence(delta_c):
    '''Check if the condition for convergence is met'''

    if delta_c < epsilon:
        return 1
    return 0


def distance_pp(p1, p2, d):
    '''Calculate the euclidean distance between two points in R^d'''

    return math.sqrt(sum([(p2[i]-p1[i])**2 for i in range(d)]))


def argmin_distance(centroids, p, k, d):
    '''Input: centroids-list of centroids [k*d], p-point in R^d. 
    Maintaining the minimum index from p to the centroids.'''

    min_index = 0
    d_min = distance_pp(centroids[0],p, d)
    for i in range(k):
        d_cur = distance_pp(centroids[i],p, d)
        if d_cur<d_min:
            min_index = i
            d_min = d_cur
    return min_index


def main():
    
    points = []
    centroids = []

    file_name = get_sys_args()

    # ------------------- for debug only - cmd:-----------------
    # python .\kmeans.py 3 100 input_1.txt
    # python .\kmeans.py 7 input_2.txt
    # python .\kmeans.py 15 300 input_3.txt
    # ----------------------------------------------------------    

    read_file(file_name, points)
    
    # check k and iter validity
    n = len(points) 
    d = len(points[0])
    if k<2 or n <= k:
        print("Invalid number of clusters!")
        sys.exit()
    if iterations < 2 or 999 < iterations:
        print("Invalid maximum iteration!")
        sys.exit()

    centroids, points = create_centroids(centroids, points, k, d)

    kmeans(points, centroids, n, k, d)
    print_2d_mat(centroids)

# Check if this module is being run as the main module
if __name__ == '__main__':
    # Call the main function
    main()

     

File: kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/*---Global Variables---*/
int n = 0;
int k = 0;
int d = 0;
int iter = 200; /* defult maximum iteration */

enum boolean {
    FALSE,    /* 0*/
    TRUE  /* 1*/
};

enum errmsg {
    NO_ERROR,       /* 0*/
    GEN_ERR,        /* 1*/
    INVAL_CLUST,    /* 2*/
    INVAL_ITER      /* 3*/
};

/*---------------func prototype:--------------------*/
int isNaturalNumber(char* str);
int readInput(int *n,int *d, double ***points);
int initCentroids(int k, int d, double ***centroids, double **points);
int initPointsCentroidsIndex(int n, int k, int **pointsCentroidsIndex);
void assign(double **points, double **centroides, int **pointsCentroidsIndex);
int update_centroids(double **points, double ***centroides, double ***prevCentroids, int *pointsCentroidsIndex);
double convergence(double **centroides, double **prevCentroids);
double distance_pp(double *p1, double *p2);
int argmin_distance(double *point, double **centroids);
void vec_sum(double *v1, double *v2);
void print_matrix_2d(int n, int m, double **matrix_2d);
double max(double num1, double num2);


int main(int argc, char *argv[])
{
    double epsilon = 0.001;
    int i = 0;
    enum errmsg err = NO_ERROR;
    /*--- define pointers for multidimensional arrays ----*/
    double **points = NULL;
    double **centroids = NULL;
    double **prevCentroids = NULL;
    int *pointsCentroidsIndex = NULL;
    
    /*--- read input from argv ---*/
    /* check if no arguments have been passed to main*/
    if(argc == 1){
        printf("An Error Has Occurred");
        return 1;
    }
    /*check if argv[1] not natural number*/
    if(isNaturalNumber(argv[1]) == FALSE){
        printf("Invalid number of clusters!");
        return 1;
    }
    k = atoi(argv[1]); /* number of clusters (centroids) */
    if(argc > 2){ /* at least 2 arguments have been passed to main */
        /*check if argv[2] is not natural number*/
        if(isNaturalNumber(argv[2]) == FALSE){
            printf("Invalid maximum iteration!");
            return 1;
        }
        iter = atoi(argv[2]);/* maximum iteration */
    }

    /*--- input check: ---*/
    if(k <= 1){
        printf("Invalid number of clusters!");
        return 1;
    }
    if(iter <= 1 || 1000 <= iter){
        printf("Invalid maximum iteration!");
        return 1;
    }

    /*--- init routine: ---*/
    err = readInput(&n, &d, &points);

    if(err == NO_ERROR && n <= k)
        err = INVAL_CLUST;
    
    if(err == NO_ERROR)
        err = initCentroids(k, d, &centroids, points);

    if(err == NO_ERROR)
        err = initCentroids(k, d, &prevCentroids, points);

    if(err == NO_ERROR)
        err = initPointsCentroidsIndex(n, k, &pointsCentroidsIndex);
    
    if(err == NO_ERROR){
    /*--- kmeans algo: ---*/
        for(i = 0; i < iter; i++)
        {
            assign(points, centroids, &pointsCentroidsIndex);

            err = update_centroids(points, &centroids, &prevCentroids, pointsCentroidsIndex);
        
            if(err != NO_ERROR || convergence(centroids, prevCentroids) < epsilon){
                break;
            }
        }
    }
    if(err == NO_ERROR){
        print_matrix_2d(k, d, centroids);
        free(*points);
        free(*centroids);
        free(*prevCentroids);
    }
    if(err == GEN_ERR)
        printf("An Error Has Occurred\n");
    if(err == INVAL_CLUST){
        printf("Invalid number of clusters!");
        free(*points);
    }

    /*--- deallocate memory ---*/
    
    free(points);
    free(centroids);
    free(prevCentroids);
    free(pointsCentroidsIndex);
    if(err == NO_ERROR)
        return 0;
    return 1;
}

int isNaturalNumber(char *str) {
    if (str == NULL || *str == '\0') {
        return FALSE; /* empty string*/
    }
    while (*str) {
        if ((*str) < '0' || '9' < (*str) ) {
            return FALSE; /* Not a natural number*/
        }
        str++;
    }
    return TRUE; /* Valid natural number*/
}

int readInput(int* n,int* d, double*** points)
{
    int indexOfPoints_nxd = 0;
    int i = 0;
    char str[200] = {0};
    enum boolean isFirstLine = TRUE;
    /*--------for daynamic allocation--------*/
    size_t sizeOfPoints_nxd = 100;
    size_t sizeOfpoints = 100;  
    double* points_nxd = (double*) calloc(sizeOfPoints_nxd, sizeof(double));
    *points = (double**) calloc(sizeOfpoints, sizeof(double*));
    /*--------------------------------------- */
    if(points_nxd == NULL || (*points) == NULL){
        free(points_nxd);
        free(*points);
        return 1;
    }
    /* get data from stdin and write it daynamicly to points_nxd*/
    str[i] = getchar();
        do{
        if(str[i] == ',' || str[i] == '\n'){
            if(isFirstLine == TRUE && str[i] == '\n'){
                *d = indexOfPoints_nxd + 1;
                isFirstLine = FALSE;
            }
            str[i] = '\0';
            points_nxd[indexOfPoints_nxd] = strtod(str, NULL); /* str to double*/
            indexOfPoints_nxd++;
            i = 0;
        }
        else
            i++;
        if(indexOfPoints_nxd == (int) sizeOfPoints_nxd){
            sizeOfPoints_nxd *= 2;
            points_nxd = (double*)realloc(points_nxd, sizeof(double) *  sizeOfPoints_nxd);
            if(points_nxd == NULL){
                free(*points);
                return 1; /* alloc problem running*/
            }
        }
    str[i] = getchar();
    /* end while condition: (str[i] != 'z') use to demonstrate end of file when debuging*/
    } while(str[i] != EOF && (str[i] != 'z'));

    /* arrange the data in points as an 2 dim array, write daynamicly*/
    *n = indexOfPoints_nxd/(*d);
    *points[0] = points_nxd;
    for(i=0;i<(*n);i++){
        if(i==(int) sizeOfpoints){
            sizeOfpoints *= 2;
            *points = (double**)realloc(*points, sizeof(double*) *  sizeOfpoints);
            if((*points) == NULL){
                free(points_nxd);
                return 1;
            }
        }
        (*points)[i] = points_nxd +i*(*d);
    }
    return 0; /* successful running*/
}

int initCentroids(int k, int d, double*** centroids, double** points){
    int i = 0;
    double* centroides_kxd = (double*) calloc(k*d, sizeof(double));
    *centroids = (double**) calloc(k, sizeof(double*));
    if(centroides_kxd == NULL || (*centroids) == NULL)
        return 1;
    for(i=0;i<k*d;i++){
        centroides_kxd[i] = (*points)[i];
    }
    for(i=0;i<k;i++){
        (*centroids)[i] = centroides_kxd +i*d;
    }
    return 0; /* successful running*/
}

int initPointsCentroidsIndex(int n, int k, int** pointsCentroidsIndex){
    int i = 0;
    *pointsCentroidsIndex = (int*) calloc(n, sizeof(int));
    if((*pointsCentroidsIndex) == NULL)
        return 1;
    for(i = 0; i < k; i++){
        (*pointsCentroidsIndex)[i] = i;
    }
    return 0; /* successful running*/
}

/*Assign every point to the closest cluster*/
void assign(double **points, double **centroides, int **pointsCentroidsIndex)
{
    int i = 0;
    int min_index = 0;
    for(i = 0; i < n; i++){
        min_index = argmin_distance(points[i], centroides);
        (*pointsCentroidsIndex)[i] = min_index;
    }
}

int update_centroids(double **points, double ***centroides, double ***prevCentroids, int *pointsCentroidsIndex)
{
    int i = 0;
    int j = 0;
    int min_index = 0;
    double **temp = NULL;
    int *c_size_arr = (int*) calloc(k, sizeof(int)); /* counts the number of points in every cluster*/
    if(c_size_arr == NULL)
        return 1;

    /*--- swap operation: centroides & prevCentroids ---*/
    temp = *prevCentroids;
    *prevCentroids = *centroides;
    *centroides = temp;

    /*--- set centroides to zeros ---*/
    for(i = 0; i < k; i++)
        for(j = 0; j < d; j++)
            (*centroides)[i][j] = 0.0;

    /*--- sum up every cooordinate in every centroid ---*/
    for(i = 0; i < n; i++)
    {
        min_index = pointsCentroidsIndex[i];
        vec_sum((*centroides)[min_index], points[i]);
        c_size_arr[min_index] += 1;
    }

    /*--- calculate the new C.G for every centroid ---*/
    for(i = 0; i < k; i++)
    {
        for(j = 0; j < d; j++)
        {
            (*centroides)[i][j] = (*centroides)[i][j] / c_size_arr[i];
        }
    }
    
    free(c_size_arr);
    return 0;
}


double convergence(double **centroides, double **prevCentroids)
{
    double delta_c = 0;
    int i = 0;
    /*Check if the condition for convergence is met*/
    for(i = 0; i < k; i++)
        delta_c = max(distance_pp(centroides[i],prevCentroids[i]),delta_c);
    return delta_c;
}



double distance_pp(double *p1, double *p2)
{
    /*Calculate the euclidean distance between two points in R^d*/
    int i = 0;
    double dist= 0.0;
    for(i = 0; i < d; i++)
    {
        dist += pow((p1[i] - p2[i]) , 2);
    }
    dist = sqrt(dist);
    return dist;
}


int argmin_distance(double *point, double **centroids)
{
    /*Input: c_arr-list of centroids [k*d], p-point in R^d. 
    Maintaining the minimum index from p to the centroids*/
    int i = 0;
    int min_index = 0;
    double dist_min = distance_pp(centroids[0], point);
    double dist_cur = 0;
    for(i = 0; i < k; i++)
    {
        dist_cur = distance_pp(centroids[i], point);
        if (dist_cur < dist_min)
        {
            min_index = i;
            dist_min = dist_cur;
        }
    }
    return min_index;
}


void vec_sum(double *v1, double *v2)
{
    int i = 0;
    /*Sum v2 into v1, each component seperately*/
    for(i = 0; i < d; i++)
    {
        v1[i] += v2[i];
    }
}

void print_matrix_2d(int n, int m, double **matrix_2d)
{
    int i = 0;
    int j = 0;
    for( i=0;i<n;i++){
        for(j=0;j<m;j++){
            printf("%.4f", matrix_2d[i][j]);
            if(j<m-1)
                printf(",");
        }
        printf("\n");
    }
}

double max(double num1, double num2){
    if(num1 < num2)
        return num2;
    return num1;
}